zkvm/src/ops.rs

```diff
@@ -399,4 +407,61 @@ impl Program {
         self.0.push(Instruction::Cloak(m, n));
         self
     }
+
+    /// Takes predicate and closure to add choose operations for
+    /// predicate tree traversal.
+    pub fn choose_predicate<F>(
+        &mut self,
+        pred: Predicate,
+        choose_fn: F,
+    ) -> Result<&mut Program, VMError>
+    where
+        F: FnOnce(PredicateTree) -> Result<&mut PredicateTree, VMError>,
+    {
+        choose_fn(PredicateTree {
+            prog: self,
+            pred: pred,
+        })?;
+        Ok(self)
+    }
+}
+
+/// Adds data and instructions to traverse a predicate tree.
+pub struct PredicateTree<'a> {
+    prog: &'a mut Program,
+    pred: Predicate,
+}
+
+impl<'a> PredicateTree<'a> {
+    /// Left Predicate branch
+    pub fn left(&mut self) -> Result<&mut Self, VMError> {
+        let (l, r) = self.pred.to_disjunction()?;
+        self.pred = l.clone();
+
+        self.prog.push(l);
+        self.prog.push(r);
+        self.prog.left();
+
+        Ok(self)
+    }
+
+    /// Right Predicate branch
+    pub fn right(&mut self) -> Result<&mut Self, VMError> {
+        let (l, r) = self.pred.to_disjunction()?;
+        self.pred = r.clone();
+
+        self.prog.push(l);
+        self.prog.push(r);
+        self.prog.left();
+
+        Ok(self)
+    }
+
+    /// Pushes program to the stack and calls the contract protected
+    /// by the program predicate.
+    pub fn call(&mut self, prog: Program) -> &mut Self {
```

Comment:
why does it need a Program? it seems like it supposed to have this program in the `self.pred`'s witness.

Reply:
And if we move tree in and out, this would be the terminal station:

```rust
pub fn call(self) -> Result<(), ...> {
    let subprog = self.pred.to_program()?;
    self.prog.push(subprog).call();
    Ok(())
}
```